### ISSUE TITLE
Feature request - OIDC strategy: Enable using customized `state` parameter instead of always generate random string

### ISSUE THREAD

**Issue by Huan9** [AzureAD/passport-azure-ad#224]
User scenario:
1. We are using `state` parameter to store original_url before the redirection started. So once we log in user, we can redirect user back to the url user first type in. 
For example:
1. User typed 'https://sonoma.hockeyapp.com/apps/hhtest';
2. Our service found user need to log in and redirect them against AAD to log in using passport OIDC strategy.
3. Once user successful log in and grant our service permission to their profile, we need to redirect user back to 'https://sonoma.hockeyapp.com/apps/hhtest'.

Usually this can be done by encoding the `redirect url` in the `state` parameter, but our OIDC strategy always generate a new string.

I have a proposed fix here but no permission to check in, here is a quick summary of the change:

1. adding a function in `lib\stateHandler.js`:
```exports.getCustomizedState = function(req, sessionKey) {
  if(!req.session || !req.session[sessionKey] || !req.session[sessionKey].state) {
    return null;
  } else {
    return req.session[sessionKey].state;
  }
}```
2. in `oidcstrategy.js` replace L1138+ L1139 with below 3 lines:
```
// add state to params and session, use a randomly generated one if no customized state value passed in
    let state = stateHandler.getCustomizedState(req, self._key) ? stateHandler.getCustomizedState(req, self._key) : aadutils.uid(32);
    params.state = state;```

I tested locally and seems like working. I am in MS HockeyApp team, can we chat to get this issue resolved? Thanks!

**Comment by Huan9** [comment]
Or if I you can give me permission, I can prepare a PR with above change for you to review and do the fix. Thanks!

**Comment by lovemaths** [comment]
@Huan9 How about provide the customized state in the `options` of `passport.authenticate`? Will it fit your need? 

```
passport.authenticate('name', { state: 'the_state_you_want_to_use'});
```

Then in oidcStrategy.js, I can do something like the following:
```
// use customized state if it is provided, otherwise, generate a random string for state
let state = options.state? options.state : aadutils.uid(32);
params.state = state;
```

**Comment by Huan9** [comment]
Hey, thank you for the quick replay. 
I tried that before I propose the other options, this will not work. When the call hit oidcStragegy.js, the options.state will be null even I set the state value when i do authentication. Btw, I did the same thing for github oauth2, and et the session property is the only way to make it work as well.
Below is my code, as you can see the 2 line that got comment out is the only way to make things working + the proposal I made. I find your alias, if you want, we can have a debug session together through Lync. 

```
// Need to do a customize handling here to set the state, so we can redirect user back to previous pages
function customizedAADPassportAuthenticator(req: express.Request, res: express.Response, next: express.NextFunction) {
  var state = generateStateWithOriginalUrl(req.query.original_url);
  // set the passport state, have to do this manually here as we are not using the passport default function to generate the state string
  //req.session['aadoidc'] = { state: state };
  //passport.authenticate('azuread-openidconnect')(req, res, next);
  passport.authenticate( 'azuread-openidconnect',{state: state })(req, res, next);
}
// Route for log in using aad identity
router.get('/auth/aad', customizedAADPassportAuthenticator);
```

**Comment by Huan9** [comment]
@lovemaths Any updates? We kind like this feature request implemented and get in asap as we plan to use it in our PROD feature that will be shipped within 2 weeks. Thanks!

**Comment by lovemaths** [comment]
@Huan9 I will discuss with you via skype.

**Comment by lovemaths** [comment]
Available/Fixed in 3.0.0.
